oauth2_http/java/com/google/auth/oauth2/TrustBoundary.java

```diff
@@ -0,0 +1,192 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpBackOffIOExceptionHandler;
+import com.google.api.client.http.HttpBackOffUnsuccessfulResponseHandler;
+import com.google.api.client.http.HttpIOExceptionHandler;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpUnsuccessfulResponseHandler;
+import com.google.api.client.json.GenericJson;
+import com.google.api.client.json.JsonParser;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Key;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.base.MoreObjects;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Represents a trust boundary that can be used to restrict access to resources. This is an
+ * experimental feature.
+ */
+public final class TrustBoundary {
+
+  static final String TRUST_BOUNDARY_KEY = "x-allowed-locations";
+  static final String GOOGLE_AUTH_TRUST_BOUNDARY_ENABLED_ENV_VAR =
+      "GOOGLE_AUTH_TRUST_BOUNDARY_ENABLE_EXPERIMENT";
+  private static final String NO_OP_VALUE = "0x0";
+  private final String encodedLocations;
+  private final List<String> locations;
+
+  public TrustBoundary(String encodedLocations, List<String> locations) {
+    this.encodedLocations = encodedLocations;
+    this.locations =
+        locations == null
+            ? Collections.<String>emptyList()
+            : Collections.unmodifiableList(locations);
+  }
+
+  private static EnvironmentProvider environmentProvider = SystemEnvironmentProvider.getInstance();
+
+  public String getEncodedLocations() {
+    return encodedLocations;
+  }
+
+  public List<String> getLocations() {
+    return locations;
+  }
+
+  public boolean isNoOp() {
+    return NO_OP_VALUE.equals(encodedLocations);
+  }
+
+  /** Represents the JSON response from the trust boundary endpoint. */
+  public static class TrustBoundaryResponse extends GenericJson {
+    @Key("encodedLocations")
+    private String encodedLocations;
+
+    @Key("locations")
+    private List<String> locations;
+
+    public String getEncodedLocations() {
+      return encodedLocations;
+    }
+
+    public List<String> getLocations() {
+      return locations;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("encodedLocations", encodedLocations)
+          .add("locations", locations)
+          .toString();
+    }
+  }
+
+  // expose this setter only for testing purposes
+  public static void setEnvironmentProviderForTest(@Nullable EnvironmentProvider provider) {
```

Comment:
For testing helpers, can you make this package-private scope. And optionally mark this with `@VisibleForTesting`?